examples/sui-move/escrow/sources/shared_example.move

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// An escrow for atomic swap of objects using shared objects without a trusted third party.
+/// Swap via shared Escrow 
+/// 
+/// 1. A user creates a shared object requesting for a specific object ID
+/// and posting up their object ID. 
+/// 
+/// 2. The other user puts in their object which has the same object ID
+/// and the exchange is fired in the same txn.
+module escrow::shared_example {
+    use std::option::{Self, Option};
+
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    /// An object held in escrow
+    struct Exchange<T: key + store> has key, store {
+        id: UID,
+        /// owner of the escrowed object
+        sender: address,
+        /// intended recipient of the escrowed object
+        recipient: address,
+        /// ID of the object `sender` wants in exchange
+        // TODO: this is probably a bad idea if the object is mutable.
+        // that can be fixed by asking for an additional approval
+        // from `sender`, but let's keep it simple for now.
+        exchange_for: ID,
+        /// the escrowed object that we store into an option because it could already be taken
+        escrowed: Option<T>,
+    }
+
+    // Error codes
+    /// The `sender` and `recipient` of the two escrowed objects do not match
+    const EMismatchedSenderRecipient: u64 = 0;
+    /// The `exchange_for` fields of the two escrowed objects do not match
+    const EMismatchedExchangeObject: u64 = 1;
+    /// The escrow has already been exchanged or cancelled
+    const EAlreadyExchangedOrCancelled: u64 = 3;
+
+    public fun create<T: key + store>(
+        recipient: address,
+        exchange_for: ID,
+        escrowed_item: T,
+        ctx: &mut TxContext
+    ) {
+        let sender = tx_context::sender(ctx);
+        let id = object::new(ctx);
+        let escrowed = option::some(escrowed_item);
+
+        transfer::public_share_object(
+            Exchange<T> {
+                id, sender, recipient, exchange_for, escrowed
+            }
+        );
+    }
+
+    /// The `recipient` of the escrow can exchange `obj` with the escrowed item
+    public fun exchange<T: key + store, ExchangeForT: key + store>(
```

Comment:
Technically speaking, this should return the `escrowed_item` rather than transfer it to the sender (Actually I'm surprised that this didn't trigger a lint, cc @awelc), which I know is unpleasantly asymmetric, but that is the way of things...